Add EIP: Hardware and Bandwidth Recommendations (#9270)
ethereum/EIPs: EIPS/eip-7870.md

---
eip: 7870

> title: Hardware and Bandwidth Recommendations for Validators and Full Nodes

Reviewer: For the EIP title, I think we should remove "for validators and full nodes" for the following reasons:

1. The current name is too long. The max is 44 characters, this is 68. With that part it is 38 characters.
2. It's sort of implied that we're talking about validators & full nodes. But I suppose we could add recommendations for light clients too. A somewhat ambiguous title would allow us to do that.

We note that it may be possible to run a client with less than the recommended specifications, however benchmarks and decision-making will be made with respect to these recommendations.

## Motivation

Clear system specifications are crucial for:

- Ensuring meaningful benchmark comparisons across different client implementations.
- Enabling informed decision-making about protocol upgrades and their resource usage implications.
- Providing clear guidance for node operators to ensure alignment with future network requirements.

Without a shared understanding of target hardware specifications:

- Benchmark results lose significance due to inconsistent testing environments.
- Decision-making becomes challenging for implementation choices, as performance characteristics are heavily hardware-dependent.
- Network participants lack clear guidance for hardware investments.

## Specification

### Roles and Their Recommended Specifications

Node operators typically run both an **Execution Layer (EL)** client and a **Consensus Layer (CL)** client on the same machine. The specifications below assume the combined resource usage of both.

> | Node Type               | Storage   | Memory | CPU Cores / Threads | **PassMark CPU Rating** | Bandwidth Download / Upload |

Reviewer 2: Are all those numbers applied based on some research made? Or based on experience from discussions with stakers?

From Nethermind perspective those are super comfortable.

Also shall we distinguish EL and CL requirements to be treated separately? I saw a lot of setups where actually both are running on different machines.

Author: I think the average user would likely run these on the same machine? 

The rationale for them is in the linked hackmd in the comment box, we could not include all of those links so the EIP has been somewhat watered down

| ----------------------- | --------- | ------ | ------------------- | ----------------------- | --------------------------- |

> | **Full Node**           | 4 TB NVMe | 32 GB  | 4c / 8t             | ~1000 ST / 3000 MT      | 50 Mbps / 15 Mbps           |

Reviewer 3: Considering history expiry, 4TB sounds like a lot to recommend just for a full node. I see it's elaborated on below but I would either provide an estimation how long it will sustain with 2 or 4tb.

Reviewer 3: I would also make it clear that these are recommended, not minimum viable specs

Reviewer 3: Same goes for 32GB, I believe 16 is enough for different client pairs I have been running. Maybe 24 could be more realistic than 32

Reviewer 2: IMO nice to work on 16 or even 8gb but this EIP is some commitment of supported hardware and some guidelines for users on what to use - I'd prefer users to use better hardware especially cost difference (especially for RAM) is like 50-100 bucks to upgrade from 16 to 32 and it gives a lot and makes us a bit more secure from some heavy blocks especially if we consider an increase of their size.

Reviewer: > Considering history expiry, 4TB sounds like a lot to recommend just for a full node. I see it's elaborated on below but I would either provide an estimation how long it will sustain with 2 or 4tb.

Yeah, we recognize that 2TB will be do-able with history expiry, but it's uncertain when that will actually ship. I would prefer the EIP's recommendations be on the safe side, 4TB.

> I would also make it clear that these are recommended, not minimum viable specs

The word "recommendation" is used several times in this doc. It's even in the title:

https://github.com/ethereum/EIPs/blob/69d6453753bc9c2ac36ee1fc0efda193561fef6d/EIPS/eip-7870.md?plain=1#L3

> Same goes for 32GB, I believe 16 is enough for different client pairs I have been running. Maybe 24 could be more realistic than 32

Yes, this 16GB is enough for today (and what my validator has) but might not be enough 3 years from now. Once again, this EIP does not state that you cannot run a full-node with less, just that 32GB is recommended.

Reviewer 3: > Yeah, we recognize that 2TB will be do-able with history expiry, but it's uncertain when that will actually ship. I would prefer the EIP's recommendations be on the safe side, 4TB. 

Afaik it's planned for first half of this year so pretty soon. But even if it ships later on, it should still be way before we hit 2TB limit. It's good to be safe, I agree, but let's be clear it's for long term and provide an estimation of how long it will actually take to fill the drive. With current spike in hardware prices, it might be more economic to wait and upgrade later then just buying 4tb immediately. 

I understand the doc is talking about recommended specs but doesn't articulate directly that there is difference between minimum viable and recommend for long term. In this case, the word recommended can be pretty relative and subjective. For someone who only wants to run a full node at home on a low budget, I would not recommend these values. And for an infrastructure provider I would suggest maybe even something even beefier. 
I think the reasoning below is a great insight but this table or motivation above should make it clear what is the target for these recommendations. Or maybe even split the table into Minimum/Recommended as software projects do when listing requirements

Reviewer 2: Setting minimum supported hardware for me is super beneficial and already mentioned that few comments above.

Will make it clear under which setups we can't put our sign and which ones we don't believe are good enough.

Reviewer: I agree, minimum hardware requirements would be useful too, but it's a much more difficult problem. I wouldn't expect the minimum for all clients to be the same. In my opinion, minimum requirements are be best left to client teams. For example, the big advantage of Nimbus is that it works very well with low-resource hardware. I believe the current recommendations will work for all clients for at least the next three years.

Reviewer 2: Hardware will be much better in next 3 years and currently recommended will probably be much cheaper, so it will be even cheaper than some super cheap ones on which any client barely works now.

Reviewer 3: I understand it's much harder to determine what are minimum specs and also harder to sustain so I don't believe it needs to part of the table. However settling now on this much room in recommended also makes me think devs might get too comfortable with the extra performance and it might become the new minimum. I believe we are building also for people with less hardware options. 

@kamilchodola I wouldn't be so sure about significant change in the hardware getting cheaper. That assumes people need to upgrade to latest hardware, their existing nodes are not going to get stronger, only rustier. And I am not very bullish on hardware getting cheaper in general, last few years are proving otherwise and although inflation is slowing down, the progress in hardware is much smaller than it used to be, Moore's law is dead and we cannot count on getting twice as performance in couple of years. Current hardware generations offer rather incremental improvements for high prices

Author: I believe we planned to prune pre-merge data in the first half of the year, which is roughly half a terrabyte.

I got this number by computing the total size of all of the files here: https://era1.ethportal.net/

Total size: 458,497,023 kB
          = 447,751.00 MB
          = 437.26 GB
 
If we use the history growth numbers from paradigm as an approximation, this buys us roughly a year: https://www.paradigm.xyz/2024/05/how-to-raise-the-gas-limit-2 
 
For full history expiry, the storage occupied will be constant, though I think its still uncertain as to when that officially lands in most clients. If it lands in a few months, I think we can change this number for full nodes and validators -- most validators AFAIK have already updated to 4TB, ethstakers for example recommend this for new validators: https://ethstaker.org/staking-hardware

| **Local Block Builder** | 4 TB NVMe | 64 GB  | 8c / 16t            | ~3500 ST / 25000 MT     | 100 Mbps / 50 Mbps          |

*Approximate single-thread (ST) and multi-thread (MT) PassMark CPU scores. For example, a PassMark ST rating of 3500 and an MT rating of 25000 typically corresponds to upper mid-range server CPUs circa 2024–2025.*

## Rationale

### Storage

- **Recommended**: 4 TB NVMe M.2 drive with:
  - **Sequential R/W**: 7,000 MB/s
  - **Random 4K R/W**: Up to 1,000,000 IOPS
- **Why NVMe over SATA?**
  - NVMe drives have significantly higher throughput and lower latency than SATA SSDs.  
  - Drives without DRAM (DRAMless) or with QLC flash are not advised, due to lower endurance and potentially lower sustained performance.
- **On Endurance (TBW)**  
  - Running a node involves frequent writes (e.g., database updates, logs). Ensure that the SSD’s Total Bytes Written (TBW) rating is sufficient for multi-year operation.
- **Capacity Considerations**  
  - As of January 2025, 2 TB can still work, but daily chain growth makes 4 TB more future-proof.  

>   - While EIP-4444 aims to reduce historical storage requirements, the timeline for EIP-4444 remains uncertain.

Reviewer 4: ```suggestion
  - While [EIP-4444](./eip-4444.md) aims to reduce historical storage requirements, the timeline for EIP-4444 remains uncertain.
```

### Memory

> - **Why 64 GB for Validators (Attester / Local Block Builder)?**

Reviewer 4: ```suggestion
- **Why 64 GB for Validators (Attesters / Local Block Builders)?**
```

  - Running an Ethereum validator (both EL & CL clients) can be memory-intensive, with state cache dominating RAM usage.
  - Certain memory intensive client combinations have historically failed with 16 GB. It is possible to tune cache size in different clients to make it work, however we do not assume that the average user will do this. 
  - Preliminary benchmarks highlighting zk-STARK memory usage suggest cryptographic operations can demand significant RAM.
  - Relative to total hardware costs, upgrading from 32 GB to 64 GB is a small price but can improve stability and is more future proof with regards to zk-STARKs.

### CPU

> - **Single vs. Multi-thread Performance**

Reviewer 4: Maybe this section could be expanded slightly, it doesn't give a clear rationale to me

#### Full Nodes

- **Recommended**: 50 Mbps download / 15 Mbps upload.
- **Rationale**:
  - Full nodes currently participate in sampling and must track the chain tip, but are not as latency-sensitive as attesters or local block builders.
  - They can operate on lower bandwidth but risk being a slot or more behind the chain if bandwidth capacity is severely limited.

## Quick Reference Summary

- **Full Node**  
  - **Storage**: 4 TB NVMe  
  - **RAM**: 32 GB  
  - **CPU**: 4 cores / 8 threads (~1000 ST / ~3000 MT PassMark)  
  - **Bandwidth**: 50 Mbps down / 15 Mbps up  

- **Attester**  
  - **Storage**: 4 TB NVMe  
  - **RAM**: 64 GB  
  - **CPU**: 8 cores / 16 threads (~3500 ST / ~25000 MT PassMark)  
  - **Bandwidth**: 50 Mbps down / 25 Mbps up  

- **Local Block Builder**  
  - **Storage**: 4 TB NVMe  
  - **RAM**: 64 GB  
  - **CPU**: 8 cores / 16 threads (~3500 ST / ~25000 MT PassMark)  
  - **Bandwidth**: 100 Mbps down / 50 Mbps up  

## Backwards Compatibility

This EIP is informational and requires no protocol changes. We recommend that future EIPs include an assessment of their impact on these hardware recommendations.

> ## Security Considerations

Reviewer 2: One thing I'd suggest doing prior to applying that is to do any research among community either by pool on X or anything else which can have bigger outreach.

If it will appear that 51% of users have weaker than specified setup (which is super unlikely but still) we could potentially open a route for some unexpected attacks because we will test on too good hardware.

At least worth to know in case of noticed issue which affect weaker setups how much network (in terms of nodes) is put in danger.

Author: We could alternatively, test against the old NUCs for a year too though the current idea was that instead of us polling N validators, we would give them a tool to allow them to assess themselves and decide when they would want to update

- Benchmark results lose significance due to inconsistent testing environments.
- Decision-making becomes challenging for implementation choices, as performance characteristics are heavily hardware-dependent.
- Network participants lack clear guidance for hardware investments.

## Specification

### Roles and Their Recommended Specifications

Node operators typically run both an **Execution Layer (EL)** client and a **Consensus Layer (CL)** client on the same machine. The specifications below assume the combined resource usage of both.

| Node Type               | Storage       | RAM   | CPU Cores / Threads | **PassMark CPU Rating**                            | Download / Upload speed  |
|-------------------------|---------------|-------|---------------------|----------------------------------------------------|--------------------------|
| **Full Node**           | 4 TB NVMe     | 32 GB | 4c / 8t             | ~1000 ST / 3000 MT                                 | 50 Mbps / 15 Mbps        |
| **Attester**            | 4 TB NVMe     | 64 GB | 8c / 16t            | ~3500 ST / 25000 MT                                | 50 Mbps / 25 Mbps        |

> | **Local Block Builder** | 4 TB NVMe     | 64 GB | 8c / 16t            | ~3500 ST / 25000 MT                                | 100 Mbps / 50 Mbps       |

Reviewer: The table needs to be re-formatted again 😄

Reviewer: Also, I think we should change "RAM" in the title to "Memory".

Reviewer: Also "speed" in "Download / Upload speed" is lowercase, when it probably shouldn't be. I'd be in favor of renaming this title to "Bandwidth Download / Upload".

- **On Endurance (TBW)**  
  - Running a node involves frequent writes (e.g., database updates, logs). Ensure that the SSD’s Total Bytes Written (TBW) rating is sufficient for multi-year operation.
- **Capacity Considerations**  
  - As of January 2025, 2 TB can still work, but daily chain growth makes 4 TB more future-proof.  
  - While EIP-4444 aims to reduce historical storage requirements, the timeline for EIP-4444 remains uncertain.

### Memory

- **Why 64 GB for Validators (Attester / Local Block Builder)?**
  - Running an Ethereum validator (both EL & CL clients) can be memory-intensive, with state cache dominating RAM usage.
  - Certain memory intensive client combinations have historically failed with 16 GB.  
  - Preliminary benchmarks (e.g., [zk-STARK memory usage](https://hackmd.io/@han/bench-hash-in-snark)) suggest cryptographic operations can demand significant RAM.  
  - Relative to total hardware costs, upgrading from 32 GB to 64 GB is a small price but can improve stability and is more future proof with regards to zk-STARKS.

### CPU

- **Single vs. Multi-thread Performance**

>   - Attesters and local block builders benefit from both high single-thread performance and solid multi-thread performance.  

Reviewer: The word "solid" stands out here. Feels too informal. I'd probably rephrase somehow. Maybe:
```suggestion
  - Attesters and local block builders benefit from both high single-thread and multi-thread performance.  
```
Also, this statement doesn't really compare ("single vs multi") or tell why it's important. Just that it is.

  - In extreme cases, a low-bandwidth node could propose partially full blocks or one that includes fewer blobs to mitigate slow broadcast.

#### Attesters

- **Recommended**: 50 Mbps download, 25 Mbps upload.
- **Minimum tested**: 15 Mbps (as per [ethPandaOps](https://ethpandaops.io/) simulations where 40% of the network had Gigabit connections and the other 60% had 15 Mbps upload).
  - However, real-world performance depends on peer network topology. A node with poor bandwidth could in theory quickly share data with a well-connected peer with good bandwidth which means that this peer could quickly seed the network.
- **Rationale**:
  - 25 Mbps was chosen as a buffer to account for these miscellaneous factors that are harder to predict.

#### Validators Using mev-boost

- **Recommended**: 50 Mbps download / 25 Mbps upload.
- **Rationale**:
  - Most MEV-relay interactions involve fetching bundles and block headers, which can be done within typical broadband speeds.
  - While the local validator will also share the block with its peers, the relay will do the same which reduces the need for local bandwidth.

>   - There may however be cases where your validator will still build a local block, such as if no mev-relay responds or if the value of the mev reward is lower than the minimum configuration set in mev-boost. In these circumstances, the recommendations for **Local Block Builders** would be relevant.

Reviewer: ```suggestion
  - However, there may be cases where your validator will still build a local block, such as if no MEV-relay responds or if the value of the MEV reward is lower than the minimum configuration set in MEV-Boost. In these circumstances, the recommendations for **Local Block Builders** would be relevant.
```
* We're missing commas around "however" but it flows better if we move it to the front.
* We use `MEV-relay` in another point. For consistency, we should capitalize MEV in other places too.
  * The "B" in boost should be capitalized too. See: https://github.com/flashbots/mev-boost#what-is-mev-boost

Reviewer: Btw, there are other instances of "mev" -> "MEV" which need to be updated.

#### Full Nodes

- **Recommended**: 50 Mbps download / 15 Mbps upload.
- **Rationale**:
  - Full nodes currently participate in sampling and must track the chain tip, but are not as latency-sensitive as attesters or local block builders.

>   - They can operate on lower bandwidth but risk being a slot or more behind the chain if capacity is severely limited.

Reviewer: Just to be extra clear. I thought this was about system resources while skimming 🤦
```suggestion
  - They can operate on lower bandwidth but risk being a slot or more behind the chain if bandwidth capacity is severely limited.
```

## Quick Reference Summary

- **Full Node**  
  - **Storage**: 4 TB NVMe  
  - **RAM**: 32 GB  
  - **CPU**: 4 cores / 8 threads (~1000 ST / ~3000 MT PassMark)  
  - **Bandwidth**: 50 Mbps down / 15 Mbps up  

- **Attester**  
  - **Storage**: 4 TB NVMe  
  - **RAM**: 64 GB  
  - **CPU**: 8 cores / 16 threads (~3500 ST / ~25000 MT PassMark)  
  - **Bandwidth**: 50 Mbps down / 25 Mbps up  

- **Local Block Builder**  
  - **Storage**: 4 TB NVMe  
  - **RAM**: 64 GB  
  - **CPU**: 8 cores / 16 threads (~3500 ST / ~25000 MT PassMark)  

>   - **Bandwidth**: 100 Mbps down / 50 Mbps up  
> 
> 
> ## Backwards Compatibility

Reviewer: An extra blank line can be deleted.
```suggestion
  - **Bandwidth**: 100 Mbps down / 50 Mbps up  

## Backwards Compatibility
```

We note that you may be able to run a client with less than the recommended specifications, however benchmarks and decision-making will be made with respect to these recommendations.

## Motivation

Clear system specifications are crucial for:

- Ensuring meaningful benchmark comparisons across different implementations
- Enabling informed decision-making about protocol upgrades and their resource usage implications
- Providing clear guidance for node operators with respect to the future

Without a shared understanding of target hardware specifications:

- Benchmark results lose significance due to inconsistent testing environments
- Decision-making becomes challenging for implementation choices, as performance characteristics are heavily hardware-dependent
- Network participants lack clear guidance for hardware investments

## Specification

> ### Recommended hardware and bandwidth specifications

Reviewer 5: Maybe include a date here, so if users stumble upon this EIP they will know that this is recent, and not recommended specifications of 3 years ago.

Reviewer 6: We should keep this EIP up to date or explicitly replace it with another in the future

| Node type | Storage | RAM | CPU Cores | CPU Single Thread/Multithread rating | Download/Upload speed |
| -------- | -------- | -------- | --------| -------- |--------|
| Full node     | 4TB     | 32GB    | 4 cores/8 threads  | 1000 / 3000 | 50 Mbps / 15 Mbps |
| Attester      | 4TB     | 64GB    | 8 cores/16 threads | 3500 / 25000 | 50 Mbps / 25 Mbps |

> | Local block builder | 4TB     | 64GB    | 8 cores/16 threads | 3500 / 25000 |  100 Mbps / 50 Mbps |

Reviewer: Can we format this table so everything is aligned? So it's easier to read when viewing markdown source.

Reviewer: Also, for consistency with the other fields, we should add spacing around "/" in the CPU cores column. And "cores" should be lowercase in the column title.

Author: Yep, for future reference, I used: https://codebeautify.org/markdown-formatter#

- The CPU rating is a measure of how powerful a particular CPU is for single and multithreaded respectively. All numbers referenced are normalized according to PassMark.

> ### Recommended Prebuilds (Attester and Local block builder)

Reviewer 7: Even if you suggest different options, could you stick to a single one for the benchmarks in the future? 2 as max if we include RPI which otherwise is definitely non-represented.

Meaning this wouldn't just be the hardware recommendations but also the common denominator for all benchmarks given from now on. 
In that way, is significantly easier to compare cryptographic protocols rather than giving benchmarks on an M4 Max and then having no clue if is suitable for Attesters or block builders etc..

If not fixed to a machine, at least fixed to a number of cores a threads would definitely help.

Reviewer 8: It is not good idea nor practical to do benchmarks on single hardware. You can instead tune your CPU to the provided rating.

Reviewer 7: Mostly say so because depending on the CPU, it might have AVX-{} support or not. Same with custom instruction sets for asm. RPI won't support Intel's fast modular field custom instruction set for example.

These are all issues that adjusting to rating won't support. Also could help to dismiss or incentivize the implementation of such specific backends to squeeze as much perf as possible.

Finally, maybe not as big of an issue (I ignore that part) but different pipelining, cache lines and other kinds of similar properties can also influence some of the stuff we bench. Specifically when we work sometimes within the realm of `us/ns`.

This was the main reason behind the suggestion. But if there are well-known alternatives that solve that or this can also be "adjusted", please let me know to resolve the comment! cc: @chfast

Author: Added note about DRAM cache and TLC flash storage

> ### Recommended Prebuilds (Attester and Local block builder)

Reviewer 9: MOVE

#### Minisforum UM790 Pro

We recommend the Minisforum UM790 Pro with the following modifications:

- Minisforum UM790 Pro
- 8 cores
- 16 threads
- The barebones model will need to be updated to include 32GB RAM and 4TB storage
- CPU Rating
  - Single Thread: 3903
  - Multithread: 30367

> #### ASUS NUC 14 Pro

Reviewer 5: An idea, but maybe move those recommended prebuilds to an external site (as part of removing the brand names). HackMD will do fine. If there is some kind of external site which actually document and review setups that would also be great (but not aware of any).

Author: This makes sense -- I think people will want some document that has more precise suggestions. It could be a hackmd that says something like "concrete suggestions for EIP XXXX"

- 16 cores (6 performance + 8 efficiency cores + 2 Low power efficiency cores)
- 22 threads
- CPU Rating
  - Single Thread: 3520
  - Multithread: 25158

> ### Recommended Prebuild (Full node)

Reviewer: To match the order of the recommended hardware table, the full node section should be first.

Author: Since we don't want brands, I'll remove this whole section on recommended prebuilds

#### Raspberry Pi 5

- 4 core
- 4 threads
- CPU Rating
  - Single thread: 1487

>   - Multithred: 3428

Reviewer: I'm a little concerned that a RP5 would be unable to function as a full node. Can anyone confirm this works? For one, storage speeds are quite important and it only operates at PCIe 2.0 speeds (like ~500MB/s). This might be do-able, but it's not something I would recommend today. Instead, I would suggest a Rock 5 Model B which is all-around better.

Reviewer: Apparently a RP5 works. Very cool. Consider my concern withdrawn.

See: https://x.com/EthereumOnARM/status/1880258026676146618

>   - Multithred: 3428

Reviewer: ```suggestion
  - Multithread: 3428
```

## Rationale

### Storage

For storage, we recommend the following SSDs:

- Kingston KC3000
- Kingston Fury Renegade
- Samsung 990 Pro
- Seagate Firecuda 530
- Teamgroup MP44

> - WD Black SN850X

Reviewer 10: Do you have to list specific commercial models in an EIP? Can you focus of i/o metrics that are brand agnostic?

Reviewer 11: indeed i/o metrics would be more generalised and open up more broader choices for people

Author: I took the weakest SSD brand in terms of random reads and writes and used its i/o metrics.

This was the Kingston KC3000 -- all of them were roughly the same.

Link: https://www.kingston.com/unitedkingdom/en/ssd/kc3000-nvme-m2-solid-state-drive?capacity=2048gb

> - WD Black SN850X

Reviewer 10: Do you have to list specific commercial models in an EIP? Can you focus of i/o metrics that are brand agnostic?

Author: Fixed -- see other comment

> In particular, we recommend purchasing NVMe M.2 instead of SATA as NVMe has higher throughput.

Reviewer: Additionally, we should mention that we recommend NVMe drives with DRAM and TLC flash. QLC flash is slower and has a lower write endurance, meaning they may fail sooner.

Author: Added note about DRAM cache and TLC flash storage

As of January 2025, the storage requirements for all node types are dominated by the need to store block history.
A 2TB disk can be used; however, this is not recommended due to the rate of history growth and the uncertainty as to when EIP-4444 will be completed.

### RAM

> RAM/memory is dominated by state cache. As of January 2025, it is possible to run a full node with 16GB of RAM, however this has been known to not work with all combinations of EL and CL clients in the past.

Reviewer: Just a nit. I would prefer if the section header were "Memory" and replace "RAM" with "memory" in the paragraph. At the beginning, we could do "Memory (RAM) is dominated by..."

On 32GB vs 64GB; 32GB works right now, however we recommend 64GB as [preliminary benchmarks](https://hackmd.io/@han/bench-hash-in-snark) have shown that zk-STARKS can consume a significant amount of memory and the difference in cost relative to the entire hardware setup for a validator is insignificant.

### Bandwidth

- Statista states that as of January 2024:
  - The global average download for broadband is 92 Mbps and the global average upload is 43 Mbps.

>   - The global average download for mobile is 50 Mbps and 11 Mbps

Reviewer: Missing punctuation.
```suggestion
  - The global average download for mobile is 50 Mbps and 11 Mbps.
```

- GSMA report showing the state of mobile internet connectivity in 2024 shows that:

>   - The mobile upload speeds in Higher Income Countries (HIC) is about 18 Mbps

Reviewer: Missing punctuation.
```suggestion
  - The mobile upload speeds in Higher Income Countries (HIC) is about 18 Mbps.
```

  - The global average mobile download is 48 Mbps.

Although most users will be using fixed broadband as opposed to mobile broadband. We include the mobile broadband figures as a way to approximate lower bounds.

#### Local block builders

For validators that want to build blocks locally, we recommend global bandwidth figures inline with the global average for fixed broadband: 100 Mbps download and 50 Mbps upload.

> *If a block builder does not have the recommended bandwidth, we recommend that they build a block that is partially full and or one that includes less blobs.*

Reviewer: ```suggestion
*If a block builder does not have the recommended bandwidth, we recommend that they build a block that is partially full and or one that includes fewer blobs.*
```

#### Attesters

> For attesters, we recommend: 50 Mbps download and 25 Mbps upload.

Reviewer: The colon isn't really necessary here.
```suggestion
For attesters, we recommend 50 Mbps download and 25 Mbps upload.
```

> The minimum value recorded to work as of January 2025 has been ~15 Mbps. This was recorded in simulations done by Eth panda ops where 40% of the network had Gigabit connections and the other 60% had 15 Mbps upload.

Reviewer 5: ```suggestion
The minimum value recorded to work as of January 2025 has been ~15 Mbps. This was recorded in simulations done by ethPandaOps where 40% of the network had Gigabit connections and the other 60% had 15 Mbps upload.
```

Not sure if this is the desired casing, but took this from their site.

The 25 Mbps was chosen as a buffer to account for these miscellaneous factors that are harder to predict.

#### Validators using mev-boost

The recommendation for validators using mev-boost is the same as the recommendations for attesters.

#### Full nodes

For full nodes, whom want to follow the tip of the chain, we recommend 50 Mbps download and 15 Mbps upload.

Full nodes currently participate in sampling, however since they are not latency sensitive like attesters and local block builders, we can lower their recommended upload bandwidth.

The recommended download speed is kept the same, so that they are able to follow the chain at the same speed as an attester.

Note: A full node will still be able to operate at significantly lower bandwidth, however depending on how low they may be at least a slot behind the chain.

## Backwards Compatibility

This EIP is informational and requires no protocol changes. We recommend that future EIPs include an assessment of their impact on these hardware recommendations.

> ## Security Considerations

Reviewer 5: I think I'd like to see a section about "Future". What could we expect in the future? Would we need more storage? If I build a node using the recommended settings, what is the target duration that this setup is usable and needs no hardware upgrades?

Author: Off the top of my head, this is what I can think of. Kind of hard to be comprehensive and I'm sure there are nuances that are missed or mistakes:

**Full node**

- Download and upload speed goes down once we have peerDAS
- Storage will go down once we have EIP444
- Storage and RAM will go down once we have stateless (unclear how stateless clients will do state cache for re-org protection)
- CPU could go down further once we snarkify the EL block
- CPU and storage might go up depending on how far we increase the gas limit and whether we have EIP4444 and stateless

**Attesters**

- They inherit all of the changes from full nodes (though attesters sample more, so higher bandwidth)
- Download speed goes up due to increase in blobs

**Block builder**

- CPU usage goes up once we have stateless due to the need to create the stateless proofs
- Download and upload speed initially goes down to peerDAS
- Upload speed and cpu requirements goes down due to epbs due to having more time to create the execution payload
- Download and upload speed goes up due to increase in blobs
- CPU requirements increase due to the addition of heavier proofs like zkvm (probably will need a GPU)

 (One note here is that until we have zkvm proofs, you can still block build but at a lower throughput by setting the maximum number of blobs you can do in your EL based on your bandwidth requirements) 

I think the main entity that will change significantly is the block builder.

Other changes to consider:

- State network: This refers to the portal state network -- it would in theory, allow us to increase the gas limit and so it would increase builder requirements and the download speed of attesters and full nodes. 

The argument here is that if we increase the gas limit, we increase the rate of state growth and eventually no-one but centralized parties will be able to hold the state and they could deny access or censor the rest of the network. But with state network, we would always be able to retrieve the state (at a slower throughput).
  
- Delayed state root: depending on your client and how long it was taking to compute the state root, the builder cpu requirements can go down

- FOCIL; This effectively puts a lower bound on the block size and I think mostly affects block builders, in practice, I think the FOCIL limit will be negligible and won't affect the specs.

- Rainbow staking derivatives: This will add more roles -- separating the centralizing components like block building from the other components at the very least. 

It has been argued that this might give folks reason to pump up the requirements for the centralizing components even further like block building. I am unsure about how this will play out.

---
eip: 7870
title: Hardware and Bandwidth Recommendations for Validators and Full Nodes
description: System recommendations for Validators and Full nodes
author: Parithosh Jayanthi (@parithosh), Kevaundray Wedderburn (@kevaundray), Josh Rudolf (@jrudolf), Dankrad Feist (@dankrad), Justin Traglia (@jtraglia), Ignacio Hagopian (@jsign), George Kadianakis (@asn-d6)
discussions-to: https://ethereum-magicians.org/t/hardware-and-bandwidth-recommendations-for-full-nodes-and-validators/22675
status: Draft
type: Informational
created: 2025-01-26
requires: <EIP number(s)> # Only required when you reference an EIP in the `Specification` section. Otherwise, remove this field.
---

## Abstract

> This proposal specifies hardware and bandwidth recommendations for full nodes and validators. Validators are split into attesters and proposers. The resource heavy component of proposing a block is creating it and broadcasting the necessary data needed to ensure that other attesters view the block as valid. We use the "block builder" to further emphasize this.

Reviewer 12: Consider renaming "block builder" to "solo" or "local" block builder to emphasize the distinction with professional block builders.

We note that you may be able to run a client with less than the recommended specifications, however benchmarks and decision-making will be made with respects to these recommendations.

## Motivation

Clear system specifications are crucial for:

- Ensuring meaningful benchmark comparisons across different implementations

> - Enabling informed decision-making about protocol upgrades and their hardware implications

Reviewer 12: Both HW and bandwidth

```suggestion
- Enabling informed decision-making about protocol upgrades and their resource usage implications
```
